### PR TITLE
RANGER-4670: (hbase plugin) Config to support disabling column authorization for fully authorized column families

### PR DESCRIPTION
**Pull request by fateh288**
## What changes were proposed in this pull request?

A property/flag can be set via hbase service configs wherein during column level authorization, it can be used to determine if an authorization at column family can be used to determine if column authorization can be shortcircuited as an optimization (there will be audit behavior changes in this case). This is useful for improving performance especially for multiget and multiput workloads wherein thousands of columns can come in the authorization request and just checking column family level access can improve performance drastically

## How was this patch tested?

Existing test cases work

New test cases have been added for hbase. The test cases cover the scan operation bug addressed in RANGER-4977 and also covers new unit test cases for testing correctness of column authorization shortcircuiting. These minicluster cased test cases do not work on master branch (because they are ignored using junit ignore annotation -- currently being tracked at https://issues.apache.org/jira/browse/RANGER-4686 and was caused due to hadoop version upgrades) 
I have verified that the new added test cases work on my private fork.

Performance analysis:
Multiput workload command: 
`time hbase ltt -tn multitest -families cf0,cf1,cf2 -write 1000:2:5 -multiput -num_keys 1000 -num_regions_per_server 1`
Policy for (multitest,*,*) resource allowing access to the user
Execution time without service config for optimization:
25 seconds
Execution time with service config for optimization:
10 seconds

`time hbase ltt -tn multitest -families cf0,cf1,cf2 -write 10000:2:5 -multiput -num_keys 1000 -num_regions_per_server 1`
Policy for (multitest,*,*) resource allowing access to the user
Execution time without service config for optimization:
217 seconds
Execution time with service config for optimization:
83 seconds

Audit behavior change observed:
Command: 
`hbase ltt -tn multitest -families cf0,cf1 -write 3:2:1 -multiput -num_keys 1 -num_regions_per_server 1`
If service config for optimization is not enabled: 15 audits 
Screenshot 2024-12-02 at 4 44 33 PM

If service config for optimization is enabled: 2 audits (multitest/cf1 and multitest/cf2)
(calculation is not considering table level audits as it is constant for both cases)

### CODE REVIEW

File: agents-common/src/main/java/org/apache/ranger/authorization/hadoop/constants/RangerHadoopConstants.java [hunk 1]
```diff
@@ -51,6 +51,7 @@ public class RangerHadoopConstants {
 
 	public static final String  HBASE_UPDATE_RANGER_POLICIES_ON_GRANT_REVOKE_PROP 	     = "xasecure.hbase.update.xapolicies.on.grant.revoke";
 	public static final boolean HBASE_UPDATE_RANGER_POLICIES_ON_GRANT_REVOKE_DEFAULT_VALUE = true;
+	public static final String HBASE_COLUMN_AUTH_OPTIMIZATION = "ranger.plugin.hbase.column.auth.optimization";
```
Comment (mneethiraj) [inline agents-common/src/main/java/org/apache/ranger/authorization/hadoop/constants/RangerHadoopConstants.java]: `ranger.plugin.hbase.column.auth.optimization` => `ranger.plugin.hbase.column.auth.optimized`

File: hbase-agent/src/main/java/org/apache/ranger/authorization/hbase/RangerAuthorizationCoprocessor.java [hunk 2]
```diff
@@ -90,19 +91,28 @@ public class RangerAuthorizationCoprocessor implements AccessControlService.Inte
...
 	 * These are package level only for testability and aren't meant to be exposed outside via getters/setters or made available to derived classes.
 	 */
 	final HbaseFactory _factory = HbaseFactory.getInstance();
 	final HbaseUserUtils _userUtils = _factory.getUserUtils();
 	final HbaseAuthUtils _authUtils = _factory.getAuthUtils();
 	private static volatile RangerHBasePlugin hbasePlugin = null;
-	
+
+	public void setColumnAuthOptimizationEnabled(boolean enable) throws Exception {
+		if (hbasePlugin!=null) {
```
Comment (mneethiraj) [inline hbase-agent/src/main/java/org/apache/ranger/authorization/hbase/RangerAuthorizationCoprocessor.java]: Assign RangerAuthorizationCoprocessor. hbasePlugin to a local variable before accessing it:

```
RangerHBasePlugin plugn = RangerAuthorizationCoprocessor.hbasePlugin;

if (plugin != null) {
  plugin. setColumnAuthOptimizationEnabled(enable);
}
```

File: hbase-agent/src/main/java/org/apache/ranger/authorization/hbase/RangerAuthorizationCoprocessor.java [hunk 3]
```diff
@@ -491,14 +515,52 @@ ColumnFamilyAccessResult evaluateAccess(ObserverContext<?> ctx, String operation
 				session.resourceMatchingScope(RangerAccessRequest.ResourceMatchingScope.SELF);
 				session.ignoreDescendantDeny(true);
 			} else {
-				LOG.debug("evaluateAccess: columns collection not empty.  Skipping Family level check, will do finer level access check.");
+				boolean isColumnAuthShortCircuitingEnabled = hbasePlugin.getPropertyIsColumnAuthOptimizationEnabled();
```
Comment (mneethiraj) [inline hbase-agent/src/main/java/org/apache/ranger/authorization/hbase/RangerAuthorizationCoprocessor.java]: To be consistent, rename: `isColumnAuthShortCircuitingEnabled` => `isColumnAuthOptimizationEnabled`

File: hbase-agent/src/main/java/org/apache/ranger/authorization/hbase/RangerAuthorizationCoprocessor.java [hunk 4]
```diff
@@ -491,14 +515,52 @@ ColumnFamilyAccessResult evaluateAccess(ObserverContext<?> ctx, String operation
...
+				}
+				if (isColumnAuthShortCircuitingEnabled) {
+					session.column(null)
+							.buildRequest()
+							.authorize();
+
+					boolean isColumnFamilyAuthorized = session.isAuthorized();
+
+					//check if column family fully authorized i.e. no deny for columns
+					session.column(null)
```
Comment (mneethiraj) [inline hbase-agent/src/main/java/org/apache/ranger/authorization/hbase/RangerAuthorizationCoprocessor.java]: Can the second authz, starting with line 531, be skipped when `isColumnFamilyAuthorized` is `false`?

```
boolean isColumnFamilyAndDescendantsAuthorized = false;

if (session.isAuthorized()) {
  session.column(null)
         .resourceMatchingScope(RangerAccessRequest.ResourceMatchingScope.SELF_OR_DESCENDANTS)
         .ignoreDescendantDeny(false)
         .buildRequest()
         .authorize();

  isColumnFamilyAndDescendantsAuthorized = session.isAuthorized();
}

File: hbase-agent/src/main/java/org/apache/ranger/authorization/hbase/RangerAuthorizationCoprocessor.java [hunk 5]
```diff
@@ -491,14 +515,52 @@ ColumnFamilyAccessResult evaluateAccess(ObserverContext<?> ctx, String operation
...
+					session.resourceMatchingScope(RangerAccessRequest.ResourceMatchingScope.SELF)
+							.ignoreDescendantDeny(true);
+					if (LOG.isDebugEnabled()) {
+						LOG.debug("evaluateAccess: isColumnAuthShortCircuitingEnabled=true, isColumnFamilyAndDescendantsAuthorized={}",isColumnFamilyAndDescendantsAuthorized);
+					}
+					if (isColumnFamilyAndDescendantsAuthorized) {
+						familiesFullyAuthorized.add(family);
+						if (auditEvent != null) {
+							if (LOG.isDebugEnabled()) {
+								LOG.debug("evaluateAccess: isColumnAuthShortCircuitingEnabled=true, adding to familiesFullyAuthorized set");
```
Comment (mneethiraj) [inline hbase-agent/src/main/java/org/apache/ranger/authorization/hbase/RangerAuthorizationCoprocessor.java]: Include `family` in the log message, to be more useful.

```
LOG.debug("evaluateAccess: isColumnAuthOptimizationEnabled ={}, adding family {} to familiesFullyAuthorized", isColumnAuthOptimizationEnabled, family);
```

**Comment by fateh288** [comment]
@mneethiraj 
I have added performance analysis numbers for multiput workload and audits generated with/without the optimization enabled

File: hbase-agent/src/main/java/org/apache/ranger/authorization/hbase/RangerAuthorizationCoprocessor.java (hunk 2 above)
Comment (fateh288) [reply hbase-agent/src/main/java/org/apache/ranger/authorization/hbase/RangerAuthorizationCoprocessor.java]: Why is this assigning hbasePlugin to a local variable required ?
hbasePlugin is already initialized in the start()

Comment (mneethiraj) [reply hbase-agent/src/main/java/org/apache/ranger/authorization/hbase/RangerAuthorizationCoprocessor.java]: hbasePlugin is a static volatile member, whose value could change between if at line 108 and referencing at line 109. Hence copying to a local variable is a safe practice.

Comment (fateh288) [reply hbase-agent/src/main/java/org/apache/ranger/authorization/hbase/RangerAuthorizationCoprocessor.java]: Addressed in the latest commit

**Review by mneethiraj — APPROVED** [review]

**Pull request merged**